File: 0001-0500/0018/0018.4-sum.cs
/*
 * @lc app=leetcode id=18 lang=csharp
 *
 * [18] 4Sum
 */

// @lc code=start
public class Solution {
    public IList<IList<int>> FourSum(int[] nums, int target)
    {
        Array.Sort(nums);

        var originalK = 4;
        var matches = new List<IList<int>>();
        var backtrack = new int[originalK - 2];

        KSum(target, 0, originalK);

        return matches;

        void KSum(long target, int start, int k)
        {
            if (start > nums.Length - k)
            {
                return;
            }

            var avg = target / k;
            if (nums[start] > avg || avg > nums.Last())
            {
                return;
            }

            if (k == 2)
            {
                TwoSum(target, start);
                return;
            }

            for (var i = start; i < nums.Length; i++)
            {
                if (i == start || nums[i - 1] != nums[i])
                {
                    var depth = originalK - k;
                    backtrack[depth] = nums[i];

                    KSum(target - nums[i], i + 1, k - 1);
                }
            }
        }

        void TwoSum(long target, int start)
        {
            var (left, right) = (start, nums.Length - 1);

            while (left < right)
            {
                var sum = nums[left] + nums[right];

                if (sum < target)
                {
                    left++;
                    continue;
                }
                
                if (sum > target)
                {
                    --right;
                    continue;
                }

                matches.Add(new List<int>(backtrack) { nums[left], nums[right] });

                left++;
                right--;

                while (left < right && nums[left] == nums[left - 1])
                {
                    left++;
                }
            }
        }
    }
}
// @lc code=end

File: 0001-0500/0018/0018.4-sum.py
#
# @lc app=leetcode id=18 lang=python3
#
# [18] 4Sum
#

# @lc code=start
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        i = 0
        L = len(nums)
        res = []
        while i < L-3:
            if target-nums[i] < 3*nums[i+1] or target-nums[i] > 3*nums[-1]:
                    while i < L-4 and nums[i] == nums[i+1]:
                        i+=1
                    i+=1
                    continue
            j = i+1
            while j < L-2:
                if target-nums[i]-nums[j] < 2*nums[j+1] or target-nums[i]-nums[j] > 2*nums[-1]:
                    while j < L-3 and nums[j] == nums[j+1]:
                        j+=1
                    j+=1
                    continue
                left = j+1
                right = L-1
                new_target = target-nums[i]-nums[j]
                while left<right:
                    if nums[left]+nums[right] > new_target:
                        right-=1
                    elif nums[left]+nums[right] < new_target:
                        left+=1
                    else:
                        res.append([nums[i],nums[j],nums[left],nums[right]])
                        temp_left = nums[left]
                        temp_right = nums[right]
                        while(left<right and nums[left]==temp_left):
                            left+=1
                        while(left<right and nums[right]==temp_right):
                            right-=1
                while j < L-3 and nums[j] == nums[j+1]:
                    j+=1
                j+=1
            while i < L-4 and nums[i] == nums[i+1]:
                i+=1
            i+=1
        return res
# @lc code=end

